### PR TITLE
Span nesting bug

### PR DESCRIPTION
**Pull request by cliedeman**

### CODE REVIEW

File: src/__tests__/__snapshots__/integration-test.ts.snap [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+            Object {
+              "children": Array [],
+              "finished": true,
+              "id": 7,
+              "logs": Array [
+                Object {
+                  "result": "\\"2\\"",
+                },
+              ],
+              "name": "two",
```
Comment (cliedeman) [inline src/__tests__/__snapshots__/integration-test.ts.snap]: It's not very obvious from the snapshot but here we have the span `two` nested undnearth the span `one` for query 

`query {
        a {
          one
          two
        }
      }`

File: src/index.ts [hunk 2]
```diff
@@ -115,6 +115,11 @@ export default class OpentracingExtension<TContext extends SpanContext>
       childOf: parentSpan || undefined
     });
 
+    // Added to expose bug
+    console.log(
```
Comment (cliedeman) [inline src/index.ts]: I added this as an alternative to the snapshot to expose the bug:

For the query
`query {
        a {
          one
          two
        }
      }`

We get the following output

 PASS  src/__tests__/integration-test.ts
  ● Console

      Name: a Parent Name: request
      Name: one Parent Name: a
      Name: a Parent Name: request
      Name: one Parent Name: a
      Name: two Parent Name: *one* . <- This should be a

File: src/__tests__/integration-test.ts [hunk 3]
```diff
@@ -2,11 +2,81 @@ import * as express from "express";
...
+  options?: any;
+  logs: any[];
+  finished: boolean;
+}
+
+interface MockSpanTree extends MockSpan {
+  children: MockSpanTree[];
+}
+
+const buildSpanTree = (spans: MockSpan[]) => {
```
Comment (DanielMSchmidt) [inline src/__tests__/integration-test.ts]: I think an explicit return type might help new people (like me) here to understand the data structure better. But this is more of a nit-pick, no show stopper for me

File: src/__tests__/integration-test.ts [hunk 4]
```diff
@@ -2,11 +2,81 @@ import * as express from "express";
...
+
+const buildTree = (
+  parent: MockSpanTree,
+  spansByParentId: Map<number, MockSpan[]>
+) => {
+  if (spansByParentId.has(parent.id)) {
+    const spans = spansByParentId.get(parent.id);
+    spansByParentId.delete(parent.id);
+
+    // TODO: do we need to sort?
```
Comment (DanielMSchmidt) [inline src/__tests__/integration-test.ts]: I wonder if we gain anything by sorting by time. Would we ensure something for this middleware or would it make stuff too brittle for changes in apollo 🤔

File: src/__tests__/integration-test.ts [hunk 5]
```diff
@@ -92,4 +176,24 @@ describe("integration with apollo-server", () => {
...
+        a {
+          one
+          two
+        }
+      }`
+      })
+      .expect(200);
+
+    const tree = buildSpanTree(tracer.spans);
+    expect(tree).toMatchSnapshot();
```
Comment (DanielMSchmidt) [inline src/__tests__/integration-test.ts]: I love the idea of snapshot testing this!

File: src/index.ts (hunk 2 above)
Comment (DanielMSchmidt) [reply src/index.ts]: Thank you, this helps understand the issue better.

File: src/__tests__/__snapshots__/integration-test.ts.snap (hunk 1 above)
Comment (DanielMSchmidt) [reply src/__tests__/__snapshots__/integration-test.ts.snap]: This means now we would need to change the snapshot so it fails and put your new implementation on it, so the bug is fixed?

**Comment by cliedeman** [comment]
I should have mentioned this code is not merge ready :), just meant to decide on a way forward

File: src/__tests__/__snapshots__/integration-test.ts.snap (hunk 1 above)
Comment (cliedeman) [reply src/__tests__/__snapshots__/integration-test.ts.snap]: Yeah, but I think the snapshot is almost impossible to read so I feel like it adds little value. Maybe a custom snapshot serializer

**Comment by DanielMSchmidt** [comment]
I understand, I just had a few minutes and wanted to drop my 2 cents ;)

File: src/__tests__/integration-test.ts (hunk 3 above)
Comment (cliedeman) [reply src/__tests__/integration-test.ts]: This was done in a hurry. After looking at the code again I think we should actually subclass opentracing.Span with an in memory model. Puts us even closer to the real thing. I actually just wanted to store the children directly on the parent span as this means we don't have to bulid the tree at the end

Comment (DanielMSchmidt) [reply src/__tests__/integration-test.ts]: Hmm, nice idea, that might help us make it really neat. Could also be an own module, seems like something worth sharing with the community

**Comment by DanielMSchmidt** [comment]
@cliedeman Do you have an idea when you are going to resume work on this or shall I try to take over?

**Comment by DanielMSchmidt** [comment]
Work is continued in #83

**Pull request closed without merging**